contracts/tokenwrapper-handler/src/contract.rs

```diff
@@ -0,0 +1,237 @@
+        ExecuteMsg::SetResource {
+            resource_id,
+            contract_addr,
+        } => exec_set_resource(deps, info, resource_id, contract_addr),
+        ExecuteMsg::MigrateBridge { new_bridge } => migrate_bridge(deps, info, new_bridge),
+
+        /* ---  Tokenwrapper-handler specific execution entries --- */
+        // Proposal execution should be initiated when a proposal is finalized in the Bridge contract.
+        // by a relayer on the deposit's destination chain
+        ExecuteMsg::ExecuteProposal { resource_id, data } => {
+            execute_proposal(deps, info, resource_id, data)
+        }
+    }
+}
+
+fn exec_set_resource(
+    deps: DepsMut,
+    info: MessageInfo,
+    resource_id: [u8; 32],
+    contract_addr: String,
+) -> Result<Response, ContractError> {
+    let state = STATE.load(deps.storage)?;
+
+    // Validations
+    if info.sender != state.bridge_addr {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Save/update the mapping `resource_id => contract_addr`
+    let contract_addr = deps.api.addr_validate(&contract_addr)?;
+    set_resource(deps.storage, resource_id, contract_addr)?;
+
+    Ok(Response::new().add_attribute("method", "set_resource"))
+}
+
+fn migrate_bridge(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_bridge: String,
+) -> Result<Response, ContractError> {
+    let state = STATE.load(deps.storage)?;
+
+    // Validations
+    if info.sender != state.bridge_addr {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Migrage(update) the "bridge_addr" with "new_bridge"
+    let bridge_addr = deps.api.addr_validate(&new_bridge)?;
+    STATE.save(deps.storage, &State { bridge_addr })?;
+
+    Ok(Response::new().add_attribute("method", "migrate_bridge"))
+}
+
+fn execute_proposal(
+    deps: DepsMut,
+    info: MessageInfo,
+    resource_id: [u8; 32],
+    data: Vec<u8>,
+) -> Result<Response, ContractError> {
+    // Parse the (proposal)`data`.
+    let parsed_resource_id = element_encoder(&data[0..32]);
+    let base64_encoded_proposal = &data[32..];
+
+    let bridge_addr = STATE.load(deps.storage)?.bridge_addr;
+
+    // Validations
+    if info.sender != bridge_addr {
+        return Err(ContractError::Unauthorized {});
+    }
+    if parsed_resource_id != resource_id {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "Invalid resource id".to_string(),
+        }));
+    }
+    let tokenwrapper_addr = read_contract_addr(deps.storage, resource_id)?;
+    if !read_whitelist(deps.storage, tokenwrapper_addr.clone())? {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "provided tokenAddress is not whitelisted".to_string(),
+        }));
+    }
+
+    // Execute the proposal according to function signature
+    let msgs = vec![CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: tokenwrapper_addr.to_string(),
+        msg: Binary::from(base64_encoded_proposal),
```

Comment:
Shall we check whether the encoded proposal is valid or not? as in it is not just random bytes?

Reply:
Um, I had discussion about `checking the proposal data` with Drew in the past.
He said that there is no point to check the integrity or something like that, since the `proposal` is already encoded when creation.
I think he is right.
What if we check the validity of `proposal` data bytes here? 
It cannot be updated anyhow.